### DISCUSSION TITLE
[Proposal] More powerfull instantiation of generic types

### DISCUSSION THREAD

**Discussion by fgheysels** [dotnet/csharplang#6512] [Language Ideas]
When declaring a generic method that returns an instance of the generic type parameter, you can only do that if you specify the ´new()´ constraint on that generic type parameter, like this:

```csharp
public TSomething Create<TSomething>() where TSomething : new()
{
    return new TSomething();
}
```

This allows you to call the default constructor on that type parameter, but nothing more.

However, if I have a generic type parameter for which I specify that it must be a type that inherits from another type, and in that type you have some non-default constructors defined, it would be nice to be able to call those constructors as well.

For instance, it would be nice if this was possible:

```csharp
public class Foo
{
    public Foo(int x) {}
};

...

public TSomething Create<TSomething>() where TSomething : Foo
{
       return new Foo(3);
}
```

or even this:
```csharp
public class Foo
{
    protected Foo(int x, int y) {}
};

...
public class Bar : Foo
{
     protected Bar( int x, int y) : base(x, y) {}

     public TSomething Create<TSomething>() where TSomething : Foo
     {
          return new Foo(3, 7);
     }
}
```

**Comment by HaloFour** [comment]
There are proposals to expand constructor constraints to support parameters: #769

You can also now define `static` `virtual`/`abstract` members in an interface that can be used as a factory method:  #4436

The issue with extending that to constructors as you suggest is that constructors are not considered part of the contract, nor are derived classes required to declare such a constructor.  For example, the following has always been perfectly legal:

```cs
public class Bar : Foo
{
     protected Bar() : base(1, 2) {}
}
```

**Reply by Rekkonnect** [reply]
On top of that, constructors are generally not greatly designed around inheritance. Generics will only play nicely with constructors once you add either of the following:
- more options than just `new()`, for example `new(int, string)`, `new(string, long)`, etc.
  - my opinion on this is that it enforces a parameter ordering constraint that breaks several principles around good code infrastructure (having multiple places where you have to define the same constructor parameters definitely isn't ideal)
- introduce a constructor constraint model in roles, or another entity
  - roles is a new feature that has been going around a lot and provides abstractions in a less committal way (to my understanding at least), which you can check out [here](https://github.com/dotnet/csharplang/issues/5497)

But generally constructors suffer from a major issue; they are not specifically declarative. While they do initialize a new instance of an object, they perform additional operations based on the arguments they receive, which may be entirely different behavior than other constructors. Essentially, they're nameless initializer methods, which you can picture as the following:
```csharp
public CustomObject(string firstName, int age); // ctor

// equivalent to
public CustomObject(); // simple instance allocation
public Initialize(string firstName, int age); // initializer method called inside the ctor in the first example
```

In the above example, `Initialize` could do many things aside from just assigning the first name and the age. It could also allow different capabilities depending on which constructor was called (see `MemoryStream` for example).

And with the above being said, I lean more towards factory methods which are now thankfully being possible with static abstracts. Types inheriting interfaces with static abstract factory methods are constrained to offering a factory method named as requested, accepting the requested arguments.
